tests/data_generator/datagenerator.go

```diff
@@ -0,0 +1,300 @@
+		log.Fatal("error on parsing test data configuration file")
+	}
+	return config
+}
+
+func AddCohorts(cohorts []Cohort) {
+	log.Printf("Processing %d cohorts...", len(cohorts))
+	nextCohortId := tests.GetLastCohortId() + 1
+	for _, cohort := range cohorts {
+		var nrClones = 1
+		if cohort.CloneCount > 0 {
+			nrClones = cohort.CloneCount
+			log.Printf("** Clone directive found! Cloning this cohort %d times **", cohort.CloneCount)
+		}
+		for i := 0; i < nrClones; i++ {
+			AddCohort(nextCohortId, cohort)
+			nextCohortId++
+		}
+	}
+}
+
+func AddCohort(cohortId int, cohort Cohort) {
+	cohortName := fmt.Sprintf("%s (%d)", cohort.Cohort, cohortId)
+	log.Printf("Adding cohort_definition '%s'...", cohortName)
+	tests.ExecSQLStringOrFail(
+		fmt.Sprintf(
+			"INSERT into %s.cohort_definition "+
+				"(id,name,description) "+
+				"values "+
+				"(%d,'%s','%s')",
+			db.GetAtlasDB().Schema,
+			cohortId, cohortName, cohortName),
+		-1)
+	AddCohortPersonsAndObservations(cohortId, cohort)
+}
+
+func AddCohortPersonsAndObservations(cohortId int, cohort Cohort) {
+	// Add also to results.cohort table:
+	personIdsAdded := make([]int64, 0)
+	if cohort.PersonIds != nil {
+		log.Printf("Adding %d persons to cohort...", len(cohort.PersonIds))
+		for _, personId := range cohort.PersonIds {
+			AddPersonToCohort(cohortId, personId)
+		}
+		personIdsAdded = cohort.PersonIds
+	} else if cohort.NumberOfPersons > 0 {
+		// generate cohort.NumberOfPersons person records:
+		log.Printf("Adding %d persons to cohort...", cohort.NumberOfPersons)
+		nextPersonId := tests.GetLastPersonId(sourceId) + 1
+		for i := 0; i < cohort.NumberOfPersons; i++ {
+			personId := nextPersonId + int64(i)
+			AddPersonToCohort(cohortId, personId)
+			personIdsAdded = append(personIdsAdded, personId)
+		}
+	} else {
+		panic("Invalid config. A cohort must have persons")
+	}
+	// special case: we can have concepts listed in a cohort, meaning we
+	// want persons of this cohort to have observations registered
+	// for these concepts:
+	if cohort.Concepts != nil {
+		for i := range cohort.Concepts {
+			concept := &cohort.Concepts[i]
+			if concept.RatioOfPersons > 0 {
+				concept.PersonIdsOverride = tests.GetRandomSubset(personIdsAdded,
+					int(float32(len(personIdsAdded))*concept.RatioOfPersons))
+			} else {
+				if concept.PersonIds == nil {
+					// just copy over all personIds:
+					concept.PersonIdsOverride = personIdsAdded
+				}
+			}
+		}
+		// now that the personIds are copied over, we can add the concepts (and respective observations)
+		// for the persons of this cohort:
+		AddConcepts(cohort.Concepts)
+	}
+}
+
+func AddPersonToCohort(cohortId int, personId int64) {
+	AddPerson(personId)
+	tests.ExecSQLStringOrFail(
+		fmt.Sprintf(
+			"INSERT into %s.cohort "+
+				"(cohort_definition_id,subject_id) "+
+				"values "+
+				"(%d,%d)",
+			tests.GetResultsDataSourceForSourceId(sourceId).Schema,
+			cohortId, personId),
+		sourceId)
+}
+
+func AddPerson(personId int64) {
```

Comment:
Should we utilize batch inserts instead of per-individual inserts? Seems like it'd be faster.

Reply:
It is a good idea. I would like to propose we test this version first, and then make such improvements in a follow-up ticket if necessary.